### PR TITLE
feat: Adding support for AWS Batch

### PR DESCRIPTION
**Pull request by delagoya**
Adding support for AWS Batch, addresses issue #456

* Add the AWS Batch blueprint under `schedulers/terraform`
* Adds tutorial for running an example AWS Batch job in the created EKS cluster
* Added AWS Batch to Job Schedulers index page.
* Addresses issue #456

### What does this PR do?

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/awslabs/data-on-eks/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

Adds support for AWS Batch.

### Motivation

AWS Batch supports running high performance, compute intensive, and batch workloads on EKS. This blueprint provides a template to stand up all required AWS resources and Kubernetes permissions to enable AWS Batch managed services to launch jobs as pods on an EKS cluster. 

### More

- [X] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [X] Mandatory for new blueprints. Yes, I have added a example to support my blueprint PR
- [X] Mandatory for new blueprints. Yes, I have updated the `website/docs` or `website/blog` section for this feature
- [X] Yes, I ran `pre-commit run -a` with this PR. Link for installing [pre-commit](https://pre-commit.com/) locally

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: schedulers/terraform/aws-batch-eks/data-providers.tf [hunk 1]
```diff
@@ -0,0 +1,2 @@
+data "aws_caller_identity" "current" {}
```
Comment (vara-bonthu) [inline schedulers/terraform/aws-batch-eks/data-providers.tf:1]: Let's move these lines to `main.tf`. We don't need a separate file for this.

File: schedulers/terraform/aws-batch-eks/eks.tf [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+                    "accelerated_compute_metrics" : false, "enhanced_container_insights" : false
+                  }
+                }
+              },
+              "containerLogs" : {
+                "enabled" : true
+              }
+            }
+          },
+          "tolerations" : [{
```
Comment (vara-bonthu) [inline schedulers/terraform/aws-batch-eks/eks.tf:84]: Let's move these cluster add-ons to `addons.tf` file and leverage EKS Blueprints Add-ons TF module. 
See this example: https://github.com/awslabs/data-on-eks/blob/ff145308263203072229d4db79285919a21af42b/analytics/terraform/spark-k8s-operator/addons.tf#L363C1-L389C1

We want to make sure all the blueprints are consistently using the same approach.

File: schedulers/terraform/aws-batch-eks/eks.tf [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+      )
+    }
+  }
+  tags       = local.tags
+  depends_on = [module.vpc]
+}
+
+
+
+###############################################################################
```
Comment (vara-bonthu) [inline schedulers/terraform/aws-batch-eks/eks.tf:98]: Move the code from this line to end of the line to a batch.tf file. Anything that is required for batch.tf. file. I would also rename the file batch.tf to aws-batch.tf.

File: schedulers/terraform/aws-batch-eks/eks.tf [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+  ]
+  depends_on = [
+    module.eks,
+    kubernetes_namespace.doeks_batch_namespace,
+    kubernetes_cluster_role.batch_cluster_role,
+    kubernetes_cluster_role_binding.batch_cluster_role_binding,
+    kubernetes_role.batch_compute_env_role,
+    kubernetes_role_binding.batch_compute_env_role_binding
+  ]
+}
```
Comment (vara-bonthu) [inline schedulers/terraform/aws-batch-eks/eks.tf:249]: You dont need this. you can pass the additional auth roles through variables. Check the EKS TF module for input variables.

File: schedulers/terraform/aws-batch-eks/locals.tf [hunk 5]
```diff
@@ -0,0 +1,8 @@
+locals {
+
+  azs = slice(data.aws_availability_zones.available.names, 0, var.num_azs)
+
+  tags = merge(var.tags, {
+    Blueprint = var.eks_cluster_name
+  })
+}
```
Comment (vara-bonthu) [inline schedulers/terraform/aws-batch-eks/locals.tf:8]: Move this content to main.tf and remove this file please

File: schedulers/terraform/aws-batch-eks/outputs.tf [hunk 6]
```diff
@@ -0,0 +1,25 @@
+output "configure_kubectl_cmd" {
+  description = "Configure kubectl: make sure you're logged in with the correct AWS profile and run the following command to update your kubeconfig"
+  value       = "aws eks --region ${var.aws_region} update-kubeconfig --name ${module.eks.cluster_name}"
+  # display name
```
Comment (vara-bonthu) [inline schedulers/terraform/aws-batch-eks/outputs.tf:4]: remove the comment

File: schedulers/terraform/aws-batch-eks/provider.tf [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    host                   = module.eks.cluster_endpoint
+    cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
+    exec {
+      api_version = "client.authentication.k8s.io/v1beta1"
+      command     = "aws"
+      # This requires the awscli to be installed locally where Terraform is executed
+      args = ["eks", "get-token", "--cluster-name", module.eks.cluster_name]
+    }
+  }
+}
```
Comment (vara-bonthu) [inline schedulers/terraform/aws-batch-eks/provider.tf:28]: move this content to `main.tf` file and remove this file

File: schedulers/terraform/aws-batch-eks/variables.tf [hunk 8]
```diff
@@ -0,0 +1,105 @@
+variable "eks_cluster_name" {
+  description = "Name of the VPC and EKS Cluster"
+  default     = "doeks-batch"
```
Comment (vara-bonthu) [inline schedulers/terraform/aws-batch-eks/variables.tf:3]: shall we call this as `doeks-aws-batch`?

File: schedulers/terraform/aws-batch-eks/variables.tf [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+variable "num_azs" {
+  description = "The number of Availability Zones to deploy subnets to. Must be 2 or more"
+  default     = 2
+  type        = number
+  validation {
+    condition     = var.num_azs >= 2
+    error_message = "The number of Availability Zones must be 2 or more."
+  }
+}
+variable "eks_cluster_version" {
```
Comment (vara-bonthu) [inline schedulers/terraform/aws-batch-eks/variables.tf:22]: add empty line between each variable

File: schedulers/terraform/aws-batch-eks/variables.tf [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+  type        = bool
+  default     = true
+}
+
+variable "eks_private_cluster_endpoint" {
+  description = "Whether to have a private cluster endpoint for the EKS cluster."
+  type        = bool
+  default     = true
+}
+
```
Comment (vara-bonthu) [inline schedulers/terraform/aws-batch-eks/variables.tf:64]: We might need all of this. Let's refer to existing blueprints. You can use this example https://github.com/awslabs/data-on-eks/blob/ff145308263203072229d4db79285919a21af42b/analytics/terraform/spark-k8s-operator/vpc.tf#L7C1-L41C2

**Review by vara-bonthu — COMMENTED** [review]
Thanks for the PR @delagoya ! I did first round of review and left some comments.

File: schedulers/terraform/aws-batch-eks/eks.tf (hunk 4 above)
Comment (delagoya) [reply schedulers/terraform/aws-batch-eks/eks.tf:249]: Can you point me at an example? I am not seeing the input for config maps from the [EKS TF module](https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/latest?tab=inputs). I only see how to do this via the [aws-auth submodule](https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/latest/submodules/aws-auth) which I am using in the above code.

File: schedulers/terraform/aws-batch-eks/variables.tf (hunk 10 above)
Comment (delagoya) [reply schedulers/terraform/aws-batch-eks/variables.tf:65]: Which part are you referring to? The private vs public endpoint? The CIDR specifications?

**Comment by delagoya** [comment]
Any update on review?

**Review by vara-bonthu — APPROVED** [review]
/lgtm

**Pull request merged**